parse tuple syntax in function constructors and comma-separated arguments in `EXCEPT` assignments

johnyf opened the pull request:
These changes extend the parser to:
- recognize tuples in declarations within function constructors, for example `[<<m, n>> \in A \X B |-> m]`
- recognize comma-separated arguments in `EXCEPT` assignments, for example `[1 EXCEPT ![1, 2] = 1]`

damiendoligez reviewed (approved):
The code looks good to me. You could add a comment to describe the translation from tuple form to `LET...IN`.

johnyf commented:
Thank you for reviewing the changes and approving this pull request, and the suggestion. I added comments that describe the translation of tuple declarations using `LET...IN`. I pushed to this pull request's branch, and intend to merge.

I also added comments to the other cases within the parser for expressions delimited by square brackets and square brackets followed by a subscript.

damiendoligez commented on src/expr/e_parser.ml:
```diff
@@ -594,6 +672,175 @@ and boundeds b = lazy begin
+        *)
+        let (bounds, letin) = List.split bss in
+        (* Flatten each list of lists into a list.
+        At this point we return a list of bounds that
```
This sentence is unfinished.

johnyf replied:
Thank you for noting this, I completed the comment.

damiendoligez reviewed (commented):
Thanks a lot for the comments! They make things much clearer and allowed me to find what I think is a bug.

For the next time, I think you don't need to go into so much detail, stopping at a level of detail just above the step-by-step explanations would be fine.

johnyf commented on src/expr/e_parser.ml:
```diff
@@ -594,6 +672,175 @@ and boundeds b = lazy begin
+
+        In each function constructor, the concatenation of identifiers
+        (like "xy" above) is unique, because the TLA+ syntax ensures that
+        each identifier is unique within its context. Therefore, each bounded
```
Thank you for finding this error. Indeed, this would result in two identifiers called `fcnbnd#xyz`. This error is demonstrated in the output:

```
ASSUME g ==
         [fcnbnd#ab \in {<<1>>},
            fcnbnd#ab_1 \in {2} \X {3} |->
            LET ab == fcnbnd#ab_1[1]
                a == fcnbnd#ab_1[1]
                b == fcnbnd#ab_1[2]
            IN  ab]
PROVE  g[<<1>>, <<2, 3>>] = 1
```

that is generated for `g == [<<ab>> \in {<<1>>}, <<a, b>> \in {2} \X {3} |-> ab]`, from a new commented test that would catch this bug if the backends supported the expressions involved in the test.

I believe that I now have corrected the error, by using only the first identifier from the tuple declaration, as suggested.

johnyf commented:
Thank you for reviewing the changes.

Also, orthogonally to this pull request's topic, but since it is in the same thematic area (parsing), there is now a custom exception within the module `src/pars/pco.ml` that avoids the (sole remaining) compilation warning:

```
&& ocamlopt -annot -g -warn-error +1+2+5+6+8+10..26+29..31+36  -I . -I backend -I frontend -I expr -I module -I proof -I util -I pars -I typesystem -I smt -I ../backend -I ../frontend -I ../expr -I ../module -I ../proof -I ../util -I ../pars -I ../typesystem -I ../smt -I .. -c `basename pars/pco.ml`
File "pco.ml", line 736, characters 24-36:
736 |               | Failure "reduce_one" -> fail "incomplete expression"
                              ^^^^^^^^^^^^
Warning 52 [fragile-literal-pattern]: Code should not depend on the actual values of
this constructor's arguments. They are only for information
and may change in future versions. (See manual section 9.5)
```

This warning has been generated by OCaml version 4.12.0. As far as I can tell, there are no warnings remaining.

Upon merging, I intend to rebase the commits in this pull request that add the new functionality, comments, and amend the error, to have those in a single commit for posterity.

johnyf commented:
Rebased to squash atop of the mainline branch, as commits 5958dfaca70b07613603baadff57926aff7eba66 and 3f0fad001efa79948fbdca7000118489648791aa.
